src/class-post-republisher.php

```diff
@@ -0,0 +1,329 @@
+				'exclude_from_search'       => false,
+				'show_in_admin_all_list'    => false,
+				'show_in_admin_status_list' => false,
+			],
+		];
+
+		foreach ( $custom_post_statuses as $custom_post_status => $options ) {
+			\register_post_status( $custom_post_status, $options );
+		}
+	}
+
+	/**
+	 * Changes the post copy status.
+	 *
+	 * Runs on the `wp_insert_post_data` hook in `wp_insert_post()` when
+	 * submitting the post copy.
+	 *
+	 * @param array $data    An array of slashed, sanitized, and processed post data.
+	 * @param array $postarr An array of sanitized (and slashed) but otherwise unmodified post data.
+	 *
+	 * @return array An array of slashed, sanitized, and processed attachment post data.
+	 */
+	public function change_post_copy_status( $data, $postarr ) {
+		$post = \get_post( $postarr['ID'] );
+
+		if ( ! $post || ! $this->permissions_helper->is_rewrite_and_republish_copy( $post ) ) {
+			return $data;
+		}
+
+		if ( $data['post_status'] === 'publish' ) {
+			$data['post_status'] = 'dp-rewrite-republish';
+		}
+
+		if ( $data['post_status'] === 'future' ) {
+			$data['post_status'] = 'dp-rewrite-schedule';
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Executes the republish request.
+	 *
+	 * @param int      $post_id   The copy's post ID.
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_request( $post_id, $post_data ) {
+		if (
+			! $this->permissions_helper->is_rewrite_and_republish_copy( $post_data )
+			|| $post_data->post_status !== 'dp-rewrite-republish'
+		) {
+			return;
+		}
+
+		$original_post_id = Utils::get_original_post_id( $post_id );
+
+		if ( ! $original_post_id ) {
+			return;
+		}
+
+		// Republish taxonomies and meta.
+		$this->republish_post_taxonomies( $post_id, $post_data );
+		$this->republish_post_meta( $post_id, $post_data );
+
+		// Republish the post.
+		$this->republish_post_elements( $post_data, $original_post_id );
+
+		// Trigger the redirect in the Classic Editor.
+		if ( $this->is_classic_editor_post_request() ) {
+			$this->redirect( $original_post_id, $post_data->ID );
+		}
+	}
+
+	/**
+	 * Republishes the original post with the passed post, when using the Block Editor.
+	 *
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_after_rest_api_request( $post_data ) {
+		$this->republish_request( $post_data->ID, $post_data );
+	}
+
+	/**
+	 * Republishes the original post with the passed post, when using the Classic Editor.
+	 *
+	 * Runs also in the Block Editor to save the custom meta data only when there
+	 * are custom meta boxes.
+	 *
+	 * @param int      $post_id   The copy's post ID.
+	 * @param \WP_Post $post_data The copy's post object.
+	 *
+	 * @return void
+	 */
+	public function republish_after_post_request( $post_id, $post_data ) {
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
+			return;
+		}
+
+		$this->republish_request( $post_id, $post_data );
```

Comment:
See above, a simple  `$post` object should be enough
The same for most (all?) occurrences of `$post_id, $post_data` as parameters later in the code